src/peft/tuners/ia3/model.py

```diff
@@ -299,3 +300,104 @@ def merge_and_unload(self):
+                    resetting_active_adapter = (
+                        list(self.peft_config.keys())[0] if len(self.peft_config) > 0 else "default"
+                    )
+                    warnings.warn(
+                        f"Adapter {adapter_name} was active which is now deleted. Setting active adapter to {resetting_active_adapter}. "
+                    )
+                    target.set_adapter(resetting_active_adapter)
+
+    def add_weighted_adapter(self, adapters, weights, adapter_name):
+        """
+        This method adds a new adapter by merging the given adapters with the given weights.
+
+        Args:
+            adapters (`list`):
+                List of adapter names to be merged.
+            weights (`list`):
+                List of weights for each adapter.
+            adapter_name (`str`):
+                Name of the new adapter.
+        """
+        if adapter_name in list(self.peft_config.keys()):
+            return
+        for adapter in adapters:
+            if adapter not in list(self.peft_config.keys()):
+                raise ValueError(f"Adapter {adapter} does not exist")
+
+        target_modules_type = type(self.peft_config[adapters[0]].target_modules)
+        new_target_modules = set() if target_modules_type == list else ""
+        feedforward_modules_type = type(self.peft_config[adapters[0]].feedforward_modules)
+        new_feedforward_modules = set() if feedforward_modules_type == list else ""
+        for adapter in adapters:
+            if type(self.peft_config[adapter].target_modules) != target_modules_type:
+                raise ValueError(
+                    "all adapter configs should follow the same target modules type. "
+                    "Combining adapters with `target_modules` type being a mix of list and string is not supported."
+                )
+            if target_modules_type == list:
+                new_target_modules |= set(self.peft_config[adapter].target_modules)
+            else:
+                new_target_modules += f"({self.peft_config[adapter].target_modules})|"
+
+            if type(self.peft_config[adapter].feedforward_modules) != feedforward_modules_type:
+                raise ValueError(
+                    "all adapter configs should follow the same feedforward modules type. "
+                    "Combining adapters with `feedforward_modules` type being a mix of list and string is not supported."
+                )
+            if feedforward_modules_type == list:
+                new_feedforward_modules |= set(self.peft_config[adapter].feedforward_modules)
+            else:
+                new_feedforward_modules += f"({self.peft_config[adapter].feedforward_modules})|"
+
+        new_target_modules = list(new_target_modules) if target_modules_type == list else new_target_modules[:-1]
+        new_feedforward_modules = (
+            list(new_feedforward_modules) if target_modules_type == list else new_feedforward_modules[:-1]
+        )
```

Comment:
The whole logic has been refactored a bit for LoRA in #993 since you created this PR:

https://github.com/huggingface/peft/blob/0c16918c347bf32ad9532823068441f5fb76197a/src/peft/tuners/lora/model.py#L499-L515

Could you please adopt those changes here for consistency? Note that the type of `target_modules` and `feedforward_modules` has been changed from list to set (str is still valid).

Reply:
Let me know if my changes are correct!